airflow/providers/google/cloud/hooks/cloud_run.py

```diff
@@ -0,0 +1,245 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""This module contains a Google Cloud Run Hook."""
+from __future__ import annotations
+
+import json
+import time
+from typing import Any, Callable, Dict, List, Sequence, Union, cast
+
+from google.api_core.client_options import ClientOptions
+from googleapiclient.discovery import build
+
+from airflow.providers.google.common.hooks.base_google import GoogleBaseHook
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+DEFAULT_CLOUD_RUN_REGION = "us-central1"
+
+
+class CloudRunJobSteps:
+    """
+    Helper class with Cloud Run job status.
+    Reference: https://cloud.google.com/run/docs/reference/rest/v1/namespaces.jobs#JobStatus
+    """
+
+    EXEC_STEP_COMPLETED = "Completed"
+    EXEC_STEP_RESOURCES_AVAILABLE = "ResourcesAvailable"
+    EXEC_STEP_STARTED = "Started"
+    AWAITING_STEPS = {EXEC_STEP_STARTED, EXEC_STEP_RESOURCES_AVAILABLE}
+    ALL_STEPS = [EXEC_STEP_COMPLETED, EXEC_STEP_RESOURCES_AVAILABLE, EXEC_STEP_STARTED]
+
+
+class _CloudRunJobExecutionController(LoggingMixin):
+    """
+    Interface for communication with Google API.
+
+    :param cloud_run: Discovery resource
+    :param project_id: The Google Cloud Project ID.
+    :param execution_id: ID of a Cloud Run Job execution.
+    :param poll_sleep: The status refresh rate for pending operations.
+    :param num_retries: Maximum number of retries in case of connection problems.
+    :param wait_until_finished: If True, wait for the end of pipeline execution
+        before exiting. If False, it only submits job and check once if
+        the job is not in terminal state.
+    """
+
+    def __init__(
+        self,
+        cloud_run: Any,
+        project_id: str,
+        region: str,
+        wait_until_finished: bool,
+        execution_id: str | None = None,
+        poll_sleep: float = 10.0,
+        num_retries: int = 0,
+        job_exec_cold_start: float = 10.0,
+    ) -> None:
+
+        super().__init__()
+        self._cloud_run = cloud_run
+        self._project_id = project_id
+        self._region = region
+        self._exec_id = execution_id
+        self._poll_sleep = poll_sleep
+        self._num_retries = num_retries
+        self._execution: dict | None = None
+        self._execution_state: str | None = None
+        self._job_exec_cold_start = job_exec_cold_start
+        self._wait_until_finished = wait_until_finished
+
+    def _fetch_execution_by_id(self):
+        """
+        Helper method to fetch the execution with the specified execution ID.
+
+        :return: the Cloud Run job execution
+        """
+        self.log.debug("Fetching information for job execution %s", self._exec_id)
+        return (
+            self._cloud_run.namespaces()
+            .executions()
+            .get(name=f"namespaces/{self._project_id}/executions/{self._exec_id}")
```

Comment:
There is no way to use the official Python client library? 
I am not a Python developer, but I would assume it can be added to any Python project, no? https://cloud.google.com/python/docs/reference/run/latest

Reply:
Cannot install the official client without breaking dependencies : https://github.com/apache/airflow/issues/29643

Would appreciate any help to solve this issue :pray: :slightly_smiling_face:

Reply:
This basically requires to upgrade all Google "properties" to the latest relesed versions of their clients. See for example https://github.com/apache/airflow/issues/27292 CC: @lwyszomi @bhirsz @VladaZakharova @mhenc @kosteev. I think there is no other way around it.

We know what needs to be done, it just needs to be done :).

Reply:
Thanks @potiuk! Which files must be changed to upgrade these google properties ? I feel that we need to change "constraints" files in the end, but I wonder how these constraints are generated.

So far my lead is modifying `airflow/providers/google/provider.yaml`. I understand that precommits and/or Github actions will take care of generating the rest.

Reply:
>Cannot install the official client without breaking dependencies : https://github.com/apache/airflow/issues/29643

https://github.com/apache/airflow/issues/29643#issuecomment-1440392537

Reply:
> My concerns are more about all the potential changes brought by Google clients upgrades. Could we have insights from the Google team to anticipate the amount of work to be done and traps to avoid?

The changes are described by each library's release notes. I doubt you will get a "single" person or team view on that - each client is developed by a different team at Google and hoping that "Google" umbrella is enough to give authoritative answer for "all things google" is a bit wishful thinking I am afraid. The only option is to get the deps upgraded one-by-one and see what needs to be done.

Reply:
FYI. There were backwards incompatible changes in 2.* versions of pretty much all python clients for google services @steren . There was a major change in how the libraries were generated, that introduced incompatible changes vs. previous versions of the librarires). Those changes often required to significantly change the code that uses the libraries (and we still have quite a few of pre 2.0 libraries of google used in google provider.

Example for bigquery client: https://cloud.google.com/python/docs/reference/bigquerystorage/latest/upgrading (but it's been the same for pretty much all other libraries - there are some ~ 60 of them, many used in Airflow from pre-2 version times). 

The general comment for those libraries is this:

> The 2.0 release of the <CLIENT> client is a significant upgrade based on a [next-gen code generator](https://github.com/googleapis/gapic-generator-python), and includes substantial interface changes. Existing code written for earlier versions of this library will likely require updates to use this version. This document describes the changes that have been made, and what you need to do to update your usage.

Reply:
So - this is not API change, it's the generator change that made backwards-incompatible changes in the libraries we've been using.

Reply:
Right, a breaking change due to the client library generation process would be reflected as a new major version of the client library.

Reply:
@elakdal no reason not to use the v2 API if recent packages bumps solved the issue 🙂

Currently away from work, I can try to migrate to using the official Cloud Run client when I get back.

If this is urgent matter, please feel free to initiate new developments and I'll be happy to contribute